src/Features/CSharp/Portable/MoveDeclarationNearReference/CSharpMoveDeclarationNearReferenceCodeRefactoringProvider.cs

```diff
@@ -12,6 +12,7 @@
 namespace Microsoft.CodeAnalysis.CSharp.MoveDeclarationNearReference
 {
     [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = PredefinedCodeRefactoringProviderNames.MoveDeclarationNearReference), Shared]
+    [Export(typeof(CSharpMoveDeclarationNearReferenceCodeRefactoringProvider))]
```

Comment:
can you clarify why you needed to do this?

Reply:
Code fixer added for the new diagnostics requires the MoveDeclarationNearReference functionality. For example, if we had `int x = Computation();` and value assignment to `x` was flagged as unused, the code fix would replace it with `_ = Computation();` or `int unused = Computation();` and then move the declaration for `x` closer to it's first reference. This functionality was already implemented by this code refactoring. I have extracted out a service `IMoveDeclarationNearRefernenceService` that can be used for such code transformation. Ideally, I would have liked to change the existing refactoring to call into this service and any new clients would also do the same. However, the current implementation of this refactoring made this difficult, it has a private [State](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/MoveDeclarationNearReference/AbstractMoveDeclarationNearReferenceCodeRefactoringProvider.State.cs,21) that gets [computed before](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/MoveDeclarationNearReference/AbstractMoveDeclarationNearReferenceCodeRefactoringProvider.cs,56) registering the refactoring and then gets [captured and passed](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/MoveDeclarationNearReference/AbstractMoveDeclarationNearReferenceCodeRefactoringProvider.cs,80) to the actual code action. So, I ended up leaving the refactoring as-is, and it itself implements the `IMoveDeclarationNearRefernenceService`. I will file a follow-up item to design/clean up the refactoring so the service and refactoring can be separate pieces.

---
In reply to: [229798809](https://github.com/dotnet/roslyn/pull/30777#discussion_r229798809) [](ancestors = 229798809)

Reply:
> I will file a follow-up item to design/clean up the refactoring so the service and refactoring can be separate pieces.

I hugely dislike this.  I've seen PR after PR take this approach, and i really wish it would stop.  The repo has 5k+ issues, and these will just fall off the map as technical debt accrues.  We should be methodical and deliberate here, even if that means doing some extra work up front.  

It's just a violation of consistent design patterns we've followed to have this sort thing going on.

Reply:
> his PR needs functionality from that refactoring, which it consumes with minimal and trivial changes, while using a separate tracking issue for larger cleanup + design work. 

First, absent information otherwise, i assume "tracking issue for larger cleanup + design work" effectively means: "an issue that has a 90%+ chance of never being addressed".

> This PR needs functionality from that refactoring, which it consumes with minimal and trivial changes

I think this is the allure.  But it leads to shortcuts and poorly architected pieces.  We don't have refactorings depend on other refactorings.  The refactoring at the end of the day represents the final feature the user interacts with throught he lightbulb.   If there is shared functionality that multiple features need, we have mechanisms for doing that.  I realize it is convenient to not have to go that route, but this continually just leads to the IDE codebase being more and more cases of just doing what is convenient versus following the patterns and practices followed elsewhere.

This is a significant form of real debt as teh codebase loses consistency and cohesiveness.  Instead of common pattern people can use to accomplish these tasks, the approach taken is just do "minimal changes" for whatever is expedient, even if things are now done in compeltely different ways all over the place.

I spent probably thousands of man hours dealing with fallout of these sorts of approaches in teh past, for example when things were converted from the previous light-bulb system to the current one.  There were massive shortcuts taken all over the place (including doing things like wholesale copy/pasting vast swaths of code), presumably with the intent to "get around to it later to clean it up".  But that "get around to it later" part almost *never* sees fruition.  It is just debt that someone else down the line ends up having to deal with.  Even after years of paying this down, there are still areas in the IDE analyzers that i'm *still* working on because we went for minimal approaches to bring things up.

Reply:
Anyways, i've made my perspective clear on this.  I think this is violation of best patterns/practices in IDE layer.  It's your guys' call on if that's ok, and if anything needs to be done there.  Arguing the point more doesn't accomplish much as it will just be circular.

Reply:
> Code fixer added for the new diagnostics requires the MoveDeclarationNearReference functionality. For example, if we had int x = Computation(); and value assignment to x was flagged as unused, the code fix would replace it with _ = Computation(); or int unused = Computation(); and then move the declaration for x closer to it's first reference. 

Also, this seems like a *totally* optional part of this work that could come in a followup PR *when* there is time to properly extract out this functionality into an appropriate service that both these features can share.

Reply:
Also, looking at this feature, it seems pretty trivial to refactor.  I'd even be willing to do the work to help here *just* so we avoid this sort of shortcut.  Unfortunatley, i cannot load/build roslyn currently due to: https://github.com/dotnet/roslyn/issues/30863   Once i can, i can help out here.

Reply:
> it seems pretty trivial to refactor

My primary concern was if there would be any perf impact on computing state twice in the refactoring (before register refactoring and then again in the code action) and not capturing state and passing it to the code action. If we feel there is no concern there, refactoring it out is trivial as you mention.

Reply:
> My primary concern was if there would be any perf impact on computing state twice in the refactoring (before register refactoring and then again in the code action) and not capturing state and passing it to the code action. If we feel there is no concern there, refactoring it out is trivial as you mention.

That's a very reasonable concern.  In general, here's the intuition that seems to apply to everything.

1. analyzers need to be *fast* and *lightweight*.  They'll be called potentially millions of times, and little savings there add up a ton (especially with allocations).
2. refactorings have a little more leeway given that they basically only need to introspect what the cursor is under.

Both have hte question of "how much to store across analysis/fix". In general, the answer should be: very little.  The more that is stored, the more that is held onto.  In practice, they should recompute what they need on demand since htey already should have been lightweight to begin with.

For a single diagnostic fix, this is fine.  For a fix-all, this is also usually fine as you'll just do the analysis again, but with the spans stored in the diagnostics.  For refactorings, recomputing is also fine as generally things are so lightweight as to not matter.  

All of this also has the benefit of our code not unintentionally keeping things alive too long.

Reply:
I've opened https://github.com/dotnet/roslyn/pull/30878 to help out here. Thanks!  I guessed at the overall shape you'd need.  I think it would be easy to rejigger as necessary if you need to tweak things.